core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -199,41 +242,137 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
     removeExecutor(executorRemoved.executorId)
   }
 
+  private def killExecutor(executorId: String, timeout: Long): Unit = {
+    logWarning(s"Removing executor $executorId with no recent heartbeats: " +
+      s"${timeout} ms exceeds timeout $executorTimeoutMs ms")
+    killExecutorThread.submit(new Runnable {
+      override def run(): Unit = Utils.tryLogNonFatalError {
+        // Note: we want to get an executor back after expiring this one,
+        // so do not simply call `sc.killExecutor` here (SPARK-8119)
+        sc.killAndReplaceExecutor(executorId)
+        // SPARK-27348: in case of the executors which are not gracefully shut down,
+        // we should remove lost executors from CoarseGrainedSchedulerBackend manually
+        // here to guarantee two things:
+        // 1) explicitly remove executor information from CoarseGrainedSchedulerBackend for
+        //    a lost executor instead of waiting for disconnect message
+        // 2) call scheduler.executorLost() underlying to fail any tasks assigned to
+        //    those executors to avoid app hang
+        sc.schedulerBackend match {
+          case backend: CoarseGrainedSchedulerBackend =>
+            val isStandalone = backend.isInstanceOf[StandaloneSchedulerBackend]
+            backend.driverEndpoint.send(RemoveExecutor(executorId,
+              ExecutorProcessLost(
+                s"Executor heartbeat timed out after ${timeout} ms",
+                causedByApp = !checkWorkerLastHeartbeat || !isStandalone)))
```

Comment:
QQ: If the executor/jvm hangs (due to interaction with the application code), we wont be able to detect that anymore, right ? Or is that handled in any other way ?

Reply:
I am not sure the precise definition about "executor/jvm hangs" you mention above. Thus, I made some assumptions to explain.

Assumptions:
(1) Executor cannot send/reply any message when "executor hangs", and JVM does not crash.
=> The executor loss will be detected by HeartbeatReceiver on driver

(2) Executor can send/reply messages when "executor hangs", and JVM does not crash.
=> Cannot be detected.

(3) JVM crashes
=> The event will be detected by driver.

Reply:
For case (1), the task will keep getting rerun on a different executors - where it will keep hanging repeatedly - if the hang is due to the user code.

You are right, (2) cannot be detected - but speculative execution might help with that if it is related to the specific node (if it is related to user code/data, it will hang indefinitely until user intervention).

For (3), we get an executor removed, and will get handled as existing (not related to this PR).

Reply:
> For case (1), the task will keep getting rerun on a different executors - where it will keep hanging repeatedly - if the hang is due to the user code.

You are right. To prevent infinite retries, we may temporarily set `causedByApp` to true until we have a mechanism to detect hanging task in executor. Does it make sense?

> You are right, (2) cannot be detected - but speculative execution might help with that if it is related to the specific node (if it is related to user code/data, it will hang indefinitely until user intervention).

As I mentioned above, we need a hanging task detector on executor to detect this.

Reply:
Updated `causedByApp` to true (its default value). Will update `causedByApp` when we have a hanging task detector in executor.